packages/ra-core/src/CoreAdminRouter.tsx

```diff
@@ -7,53 +15,82 @@ import getContext from 'recompose/getContext';
 
 import { AUTH_GET_PERMISSIONS } from './auth/types';
 import { isLoggedIn } from './reducer';
-import { userLogout } from './actions/authActions';
+import { userLogout as userLogoutAction } from './actions/authActions';
 import RoutesWithLayout from './RoutesWithLayout';
-
-const welcomeStyles = {
+import {
+    Dispatch,
+    AuthProvider,
+    AdminChildren,
+    CustomRoutes,
+    CatchAllComponent,
+    LayoutComponent,
+    LayoutProps,
+} from './types';
+
+const welcomeStyles: CSSProperties = {
     width: '50%',
     margin: '40vh',
     textAlign: 'center',
 };
 
-export class CoreAdminRouter extends Component {
-    state = { children: [] };
+interface Props extends LayoutProps {
+    appLayout?: LayoutComponent;
+    catchAll?: CatchAllComponent;
+    children: AdminChildren;
+    customRoutes?: CustomRoutes;
+    loading?: ComponentType;
+}
+
+interface EnhancedProps {
+    authProvider?: AuthProvider;
+    isLoggedIn?: boolean;
+    userLogout: Dispatch<typeof userLogoutAction>;
+}
+
+interface State {
+    children: any[];
+}
+
+export class CoreAdminRouter extends Component<Props & EnhancedProps, State> {
+    state: State = { children: [] };
 
     componentWillMount() {
         this.initializeResources(this.props);
     }
 
-    initializeResources = nextProps => {
+    initializeResources = (nextProps: Props & EnhancedProps) => {
         if (typeof nextProps.children === 'function') {
             this.initializeResourcesAsync(nextProps);
         }
     };
 
-    initializeResourcesAsync = async props => {
+    initializeResourcesAsync = async (props: Props & EnhancedProps) => {
         const { authProvider } = props;
         try {
             const permissions = await authProvider(AUTH_GET_PERMISSIONS);
             const { children } = props;
 
-            const childrenFuncResult = children(permissions);
-            if (childrenFuncResult.then) {
-                childrenFuncResult.then(resolvedChildren => {
+            if (typeof children === 'function') {
```

Comment:
no need to check here: it must be a function as initializeResourcesAsync is only called in initializeResources if it's a function.

Reply:
This makes TS happy